[1단계 - 자동차 경주 구현] 아리(양유림) 미션 제출합니다.

ulimy opened the pull request:
안녕하세요 아리입니다. 좋은 인연으로 만나뵐 수 있게 되어 영광입니다 :)
첫 코드리뷰이다보니 부족한 점이 많을 것 같아 부끄럽지만 잘 부탁드립니다!

미션을 진행하면서 궁금했던 점이 있어 질문 몇가지를 적어보겠습니다.

1. 저는 보통 미션을 진행할 때, 설계를 한 뒤 코드를 작성하는 편입니다. 하지만 TDD는 테스트를 통과하는 코드를 최소한으로만 작성하고, 이를 리팩토링 하며 코드가 발전해나가야 한다고 이해하고 있습니다. 그렇다면 설계를 미리 할 필요 없이 생각이 떠오르는 대로 리팩토링 해나가며 진행하면 되는 것일까요? 그렇게 진행한다면 나무를 보다 숲을 놓쳐 흐름을 잃지 않을까 걱정입니다. 흐름을 잃지 않는 것 또한 실력인걸까요ㅠㅠ

2. 클래스를 선언하는 위치
이번 미션에서는 기능별로 클래스를 분리해보려 했습니다. 이때, 각 클래스들은 싱글톤으로 관리되는 것이 좋을지, main 메서드가 담겨있는 Application.java에서 선언한 뒤 매개변수로 넘겨주는 것이 좋을지, 전부 static으로 선언하는 것이 좋을지, 혹은 제가 모르는 방법이 있을지 궁금합니다. 

3. RacingGame 관련 테스트 코드 작성
제가 작성한 RacingGame의 경우 run을 제외한 모든 메서드가 private으로 선언되어있고, 모두 매개변수의 전달 없이 입력 및 출력만으로 이루어져 있어 테스트 코드 구성이 어려웠습니다. InputStream, OutputStream 등을 통해 입출력을 구성하여 판별하는 방법을 학습하는 것이 좋을지, 제가 모르는 방법이 있을지 궁금합니다.

4. 향후 리팩토링 방향
MVC 패턴을 적용해보고싶었으나 실패하였습니다ㅜㅜ 앞으로 남아있는 미션의 기간동안 MVC 패턴으로 리팩토링 해보고싶은데, 기능별 분리도 부족하니 이번 미션에서는 이를 더 연습하는게 좋을지, MVC 패턴을 이용해보는 것이 좋을지 조언이 듣고싶습니다!

tco0427 committed:
test: String 클래스 요구사항1

split 메소드 테스트

tco0427 committed:
refactor: String 클래스 요구사항1 메소드 분리

split 메소드 추출

tco0427 committed:
test : String 요구사항 2

괄호를 제거한 문자열 반환

tco0427 committed:
test: String 요구사항3

String 인덱스 범위 테스트

tco0427 committed:
fix: String 요구사항2

커밋 누락된 부분 확인 후 커밋

tco0427 committed:
test : SetCollection 요구사항 1

Set 크기 확인 테스트

tco0427 committed:
test : SetCollection 요구사항 2

집합 원소 존재유무 테스트

tco0427 committed:
test : SetCollection 요구사항3

집합에 존재하는지 여부에 따라 결과값이 달라지도록 테스트

tco0427 committed:
docs(readme): 문자열 덧셈 계산기 README 작성

기능 요구사항을 토대로 `구현 기능 목록` 작성

tco0427 committed:
docs(readme): 문자열 덧셈 계산기 README 수정

빈 문자열 또는 null 입력인 경우 추가

tco0427 committed:
test : 계산기 요구사항1

빈 문자열 또는 null 입력의 경우 0 반환 여부 테스트 코드 작성.

tco0427 committed:
feat: Calculator 클래스 작성

splitAndSum 메소드 정의하여 `요구사항1` 테스트 코드 통과

tco0427 committed:
test : 계산기 요구사항2

숫자 하나만을 입력하는 경우 해당 숫자 반환 여부 테스트

tco0427 committed:
docs(readme): 문자열 덧셈 계산기 README 작성

숫자 하나만 입력하는 경우 추가 작성

tco0427 committed:
feat : 계산기 요구사항2

숫자 하나만 입력하는 경우 해당 숫자 반환 기능

tco0427 committed:
test: 계산기 요구사항3

숫자 2개를 `,`로 구분하여 합한 후 반환하는지 테스트

tco0427 committed:
feat: 계산기 요구사항3

숫자 2개인 경우 `,`로 구분하여 합하도록 splitAndSum 구현

tco0427 committed:
test: 계산기 요구사항4

구분자 `:` 를 추가하여 구분하는지 테스트

tco0427 committed:
feat : 계산기 요구사항4

구분자 : 추가

tco0427 committed:
test: 계산기 요구사항5

커스텀 구분자(`//` 와 `\n` 사이)를 추가하여 구분하는지 테스트

tco0427 committed:
feat: 계산기 요구사항5

커스텀 구분자(`//` 와 `\n` 사이)로 구분해서 합을 구하도록 splitAndSum 메소드 수정

tco0427 committed:
test : 계산기 요구사항 6

음수를 입력받은 경우 RuntimeException 발생 여부 확인

tco0427 committed:
refactor: 계산기 메소드 추출

getSum()으로 추출가능한 로직 추출하여 코드 가독성을 올린다.

tco0427 committed:
chore: 계산기 메소드 추출

getSum()으로 추출가능한 로직 추출하여 코드 가독성을 올린다.

tco0427 committed:
feat : 계산기 요구사항 6

음수가 입력되는 경우 RuntimeException 발생

tco0427 committed:
feat : 계산기 요구사항 6

음수가 입력되는 경우 RuntimeException 발생

tco0427 committed:
Merge branch 'step1' of https://github.com/tco0427/java-racingcar into step1

tco0427 committed:
refactor: StringSplitter 클래스를 통해서 문자열 분리하는 부분 추출

StringSplitter 클래스를 별도로 두어서 문자열 분리하는 책임을 위임

tco0427 committed:
refactor: StringSplitter 싱글톤 적용

StringSplitter 싱글톤 적용

tco0427 committed:
fix: StringSplitter 클래스 getInstance() 수정

외부에서 getInstance() 메소드 호출을 위해 static으로 수정

tco0427 committed:
refactor : StringSplitter를 통해 분리된 문자열 분리 기능을 활용하도록 리팩토링

StringSplitter를 통해 분리된 문자열 분리 기능을 활용하도록 리팩토링

tco0427 committed:
refactor : 입력값 검증 부분 CalculatorValidation 클래스로 분리

CalculatorValidation에서 입력값을 검증하도록 책임을 분리

tco0427 committed:
refactor : 입력값 검증 부분 CalculatorValidation 클래스로 분리

CalculatorValidation에서 입력값을 검증하도록 책임을 분리

tco0427 committed:
fix : 보다 명확한 싱글톤 패턴 적용을 위하여 final 키워드 사용

tco0427 committed:
fix : 검증 로직 수정

누락 혹은 오류가 발생한 검증 부분 수정

tco0427 committed:
style: 공백 일관성 유지

공백을 일관성 있게 수정하여 코드 스타일 유지

tco0427 committed:
docs(readme): 자동차 경주 미션 README 작성

자동차 경주 미션 '기능 요구사항'을 토대로 '구현 기능 목록' 작성

tco0427 committed:
feat: 사용자 입력 받기

자동차 이름 목록과 시도할 횟수를 입력받는 InputView 구현

tco0427 committed:
feat: Car 객체 생성

자동차 이름, 현재 전진 횟수를 속성으로 가지는 Car 클래스 정의

tco0427 committed:
feat: 자동차 경주 게임 초기화

자동차 경주 게임을 위한 자동차 이름, 전체 시도 횟수 초기 세팅

tco0427 committed:
feat: RandomUtil 클래스 정의

0~9까지 사이의 정수를 생성해 전진 여부를 판단 및 반환

tco0427 committed:
feat: 현재 진행 상황 출력 view 구현

OutputView 에서 '차 이름'과 '전진 위치'를 받아 출력하는 로직 구현

tco0427 committed:
feat: 각 자동차에 대해 게임 진행 및 출력

각 자동차에 대해서 progress()를 호출하여 게임을 진행하고 그 결과를 OutputView를 통해서 출력

tco0427 committed:
feat: 각 자동차에 대해 게임 진행 및 출력

각 자동차에 대해서 progress()를 호출하여 게임을 진행하고 그 결과를 OutputView를 통해서 출력

tco0427 committed:
feat: 시도횟수만큼 게임 진행 반복

시도 횟수만큼 게임 진행을 반복하도록 while()문을 추가

tco0427 committed:
feat: 경주 게임 우승자 선정

경주 게임을 끝내고 난 후 우승자를 선정하도록 구현

tco0427 committed:
fix: 누락된 'Collections' import 추가

tco0427 committed:
feat: 경주 게임 우승자 출력

경주 게임을 끝내고 난 후 우승자를 출력하는 OutputView 로직 추가

tco0427 committed:
feat: Application 구현

main 메소드를 포함하고 있는 Application 클래스 구현

tco0427 committed:
fix: 출력 구문 수정

출력 순서 혹은 출력 구문 수정하였습니다.

tco0427 committed:
fix: 출력 구문 수정

출력 순서 혹은 출력 구문 수정하였습니다.

tco0427 committed:
feat: 자동차 이름 길이 검증

자동차 이름이 1글자 미만 혹은 5글자 초과인 경우 예외 발생

tco0427 committed:
feat: 자동차 이름 길이 검증

자동차 이름이 1글자 미만 혹은 5글자 초과인 경우 예외 발생

tco0427 committed:
feat: 시도 횟수 검증

시도 횟수가 숫자인지 그리고 양수인지를 검증하는 코드 추가

tco0427 committed:
feat: 자동차 이름 공백 검증

자동차 이름에 공백이 들어가면 안되므로 공백이 들어가는지를 추가 검증

tco0427 committed:
feat: 자동차 이름 중복 검증

자동차 이름이 중복되면 안되므로 중복이 있는지 확인

tco0427 committed:
feat: 자동차 이름 중복 검증

자동차 이름이 중복되면 안되므로 중복이 있는지 확인

tco0427 committed:
fix: 에러 메시지를 출력하도록 수정

에러 메시지를 출력할 수 있도록 수정

tco0427 committed:
refactor: 불필요한 toString 제거

Car 클래스의 toString() 메소드가 불필요하므로 제거하였습니다.

tco0427 committed:
fix: Test 에 SuppressWarnings 추가

단위 테스트 메소드에 '한글' 이름을 사용하는 경우 아스키 문자가 아니어서 warning 을 보이므로 `SuppressWarnings` 를 이용하도록 수정

tco0427 committed:
test: 자동차 진행 테스트

자동차 한 번 진행시 '전진' 혹은 '정지' 하는지 테스트

tco0427 committed:
test: 자동차 반복 진행 테스트

자동차 반복 진행시 의도한대로 제대로 동작하는지 확인

tco0427 committed:
test: 자동차 이름 길이 테스트

자동차 이름 길이가 1글자 이상, 5글자 이하인지 확인

tco0427 committed:
test: 자동차 이름 공백 포함 테스트

자동차 이름 공백 포함 여부 테스트

tco0427 committed:
test: 자동차 이름 중복 테스트

자동차 이름이 중복되면 안되므로 이를 테스트

tco0427 committed:
test: 자동차 이름 중복 테스트

자동차 이름이 중복되면 안되므로 이를 테스트

tco0427 committed:
chore: VerificationUtilTest `DisplayName` 어노테이션 포함

tco0427 committed:
test: 시도횟수 검증 테스트

시도횟수가 숫자가 아니면 예외가 제대로 발생하는지, 음수인 경우에도 제대로 예외가 발생하는지 테스트

tco0427 committed:
fix: 누락된 static import 추가

누락된 `validateAttempt` static import 추가

ulimy committed:
refactor : 에러 메시지 상수 분리

ulimy committed:
refactor : static 키워드를 이용하여 ErrorMessage 클래스 사용 간소화

ulimy committed:
refactor : 각종 조건을 위한 값 상수분리

lalize commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    public void progress(){
+        if(checkMoveOrWait()) {
```
다른데는 다 띄우셨는데 안 띄우셔서 코멘트 남겨요~

```suggestion
        if (checkMoveOrWait()) {
```

ulimy replied:
수정 했습니다!

lalize commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Override
+    public int compareTo(Car other) {
```
Comparable 구현 👍

lalize replied:
다만 나중에 요구사항이 추가되어 Car 끼리의 비교를 position이 아닌 다른 값으로도 비교해야 하는 경우가 생기면 어떻게 될까요?
단순히 원시값을 포장하는 객체가 아닌 경우(Car 처럼) 비교하는 메서드를 따로 만드는 게 가독성도 좋고 유지보수 측면에서도 좋을 수 있어요.

ulimy replied:
제가 잘 이해한 것인지 모르겠지만 비교기준이 변경되는 경우 Car가 아닌 RacingCar에서 관리되는 것이 더 좋을 것 같아 RacingCar에 구현되도록 변경하였고, 비교하는 메서드도 따로 선언해보았습니다 :)

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+    private List<String> winners = new ArrayList<>();
+    private int totalAttemptCount;
+
+    public void start() throws IllegalArgumentException {
```
throws를 해준 이유가 있을까요?

ulimy replied:
자동완성된 코드를 삭제하는 것을 놓쳤나봅니다ㅠㅠ 수정했습니다!

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+public class RacingGame {
+
+    private List<Car> carList = new ArrayList<>();
+    private List<String> winners = new ArrayList<>();
```
우승자는 carList로 구할 수 있는데 멤버 변수로 가지고 있어야 할까요?

ulimy replied:
불필요한 필드였네요.. 수정했습니다!

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+
+public class RacingGame {
+
+    private List<Car> carList = new ArrayList<>();
```
List 대신 Set을 쓰게 된다면 carList로 된 모든 부분을 carSet으로 바꿔야겠군요 ㅠ
변수명에 타입을 적지 않아도 변수의 타입으로 어떤 변수인지 추론 가능하니 carList 대신 cars와 같이 복수형을 써주세요.

ulimy replied:
불필요한 변수명에 적힌 타입 수정하였습니다! 또한 Set이 더 적절한 자료구조인 것 같아 반영해보았습니다!

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+
+        for (String name : names) {
+            validateCarName(name);
+            carList.add(new Car(name));
```
다른 곳에서 new Car를 직접 호출하는 경우 Car의 이름이 5자를 초과해서 생성도 가능할 것 같네요.
Car의 이름은 Car의 도메인 정책이 아닌 걸까요?

ulimy replied:
각 클래스 도메인 정책은 해당 클래스 내에서 실행하는 것이 더 좋다는 것 새롭게 알게되었습니다. 수정했습니다!

lalize commented on src/main/java/racingcar/util/VerificationUtil.java:
```diff
@@ -0,0 +1,62 @@
+
+public class VerificationUtil {
+
+    private VerificationUtil() {
```
인스턴스화 막기 👍

ulimy replied:
😍

lalize commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,23 @@
+
+public class InputView {
+
+    private static final Scanner scanner = new Scanner(System.in);
```
상수는 모두 대문자가 컨벤션 입니다:)
```suggestion
    private static final Scanner SCANNER = new Scanner(System.in);
```

ulimy replied:
수정했습니다!

lalize commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,23 @@
+        System.out.println("시도할 회수는 몇회인가요?");
+
+        return scanner.nextLine();
+    }
```
get은 단순히 값을 가져오는 느낌의 네이밍인 것 같아요. 사용자 입력을 받는 것에 적합한 네이밍으로 수정하면 어떨까요?

ulimy replied:
입력을 받는 부분이니 input으로 수정해보았습니다!

lalize commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.List;
+
+public class OutputView {
```
오잉 얘는 private 기본 생성자가 없네요

ulimy replied:
놓친 부분이네요ㅜㅜ 수정했습니다!

lalize commented on src/test/java/racingcar/entity/CarTest.java:
```diff
@@ -0,0 +1,43 @@
+        assertThat(car.getPosition())
+                .isGreaterThanOrEqualTo(0)
+                .isLessThanOrEqualTo(1);
+    }
```
프로덕션 코드에 랜덤 값이 4 이상인데 전진하지 않는 버그가 숨어있는 경우 해당 테스트로 발견이 가능한가요?

ulimy replied:
발견 가능하도록 수정했습니다!

lalize commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,30 @@
+        System.out.println("실행 결과");
+    }
+
+    public static void printLine() {
```
굳이 메서드로 빼야할까요?

ulimy replied:
불필요한 메서드 선언이었네요! 수정했습니다.

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+
+        getWinners();
+        printWinners(winners);
+    }
```
비즈니스 로직과 UI 로직을 분리하고 우승자를 구하는 로직을 테스트 해주세요.

ulimy replied:
비즈니스 로직과 UI 로직 분리는 해보았으나 올바른 우승자를 선택하는지 확인하는 테스트 코드를 작성하지 못했습니다ㅠㅠ 제가 작성한 코드로는 어떻게 하면 좋을지 조언을 들을 수 있을까요? getCars 메서드를 선언하여 확인해보는 방법, 우승자 선별하는 코드를 다른 클래스로 분리하여 cars를 매개변수로 전달하도록 한 뒤 테스트 하는 방법 등을 떠올려보았으나 확신이 서지 않습니다..

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+        initRacingCarGame();
+        startMessage();
+
+        while (totalAttemptCount --> 0) {
```
```suggestion
        while (totalAttemptCount-- > 0) {
```

ulimy replied:
수정했습니다!

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,86 @@
+import static racingcar.view.InputView.getCarNames;
+import static racingcar.view.OutputView.*;
+
+public class RacingGame {
```
RacingGame은 자동차 생성, 경기 진행, 우승자 구하기 등 핵심 비지니스 로직만 갖도록 UI로직과 분리하면 좋을 것 같아요.

ulimy replied:
분리해보았습니다!

lalize commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -0,0 +1,14 @@
+
+public class RandomUtil {
+
+    public static boolean checkMoveOrWait() {
```
자동차 이동 가능 여부를 판단을 하는 도메인 로직 아닌가요? 유틸성 클래스에 있어도 될까요?

ulimy replied:
RandomUtil은 랜덤 값 생성만 진행하도록 수정했습니다!

lalize commented on src/main/java/racingcar/constant/ConditionInfo.java:
```diff
@@ -0,0 +1,9 @@
+package racingcar.constant;
+
+public class ConditionInfo {
```
모든 조건에 대한 상수를 하나의 클래스로 관리하게 되면 규모가 커짐에 따라 이해하기 어려운 코드가 될 수 있을 것 같아요.
동일한 값을 사용해야 하는 곳에서 해당 클래스의 존재를 모르고 중복 선언하여 사용할 수도 있겠죠.
실제로 해당 상수를 사용하는 객체가 상수를 갖도록 하는 건 어떨까요? 유지보수 측면에서도 분명 좋을 거라고 생각합니다 :)

ulimy replied:
상수를 객체 내에 선언하는 방법! 정말 좋은 것 같습니다. 해당 클래스 관련 정보들이 클래스 내에 선언되어있으니 한 눈에 복 훨씬 편하더라구요! 감사합니다 :)

lalize commented on src/main/java/racingcar/Application.java:
```diff
@@ -0,0 +1,16 @@
+        try {
+            racingGame.start();
+        } catch (IllegalArgumentException e) {
+            printErrorMessage(e.getMessage());
```
에러 출력(UI 로직)은 RacingGame 도메인과 분리되어 있네요 :)

ulimy replied:
UI로직 전부 RacingGame과 분리되도록 수정해보았습니다!

lalize reviewed (changes requested):
안녕하세요! 아리 🐣
첫 미션 고생 많으셨어요~ 
작성해 주신 코드에 몇 가지 코멘트를 남겼으니 확인 부탁드릴게요 :)

---

TDD 관련 질문에 대해서는 간단히 답을 드리자면 모든 요구사항을 만족하는 완벽한 설계는 없다고 생각합니다. 이미 설계한 부분도 리팩토링 대상이라고 생각하시고 TDD를 진행하시면 되지 않을까요?

나머지 질문들에 대해서는 RacingGame 도메인에서 InputView, OutputView와 같은 view에 대한 의존성을 제거하여 리팩토링을 진행하면 어느정도 해소될 것으로 보여요!

ulimy committed:
style(Car.java) : 코딩컨벤션에 맞춰 스타일 수정

ulimy committed:
style : 불필요한 throws문 제거

ulimy committed:
refactor : 불필요한 메서드 분리 통합 및 기본 생성자 작성

ulimy committed:
refactor : RacingGame 클래스가 View 클래스를 의존하지 않도록 재구성

ulimy committed:
style : 의도가 분명히 드러나도록 메서드명 수정

ulimy committed:
refactor : 도메인이 해당 클래스 내에서 관리될 수 있도록 수정

ulimy committed:
style : 분명한 의도가 드러나도록 클래스, 메서드 및 변수명 수정

ulimy committed:
refactor : 자동차 목록의 타입 List -> Set 변경

ulimy committed:
test : 입력 검증 로직 관련 테스트 작성

ulimy committed:
test : Car 클래스 관련 테스트 코드 작성

ulimy committed:
test : RacingCarGame 클래스 관련 테스트 코드 작성

ulimy committed:
style : 보다 분명한 의도가 드러나도록 클래스 및 메서드명 수정

lalize commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,44 @@
+
+    public boolean isWinner(int maxPosition) {
+        return position == maxPosition;
+    }
```
1. 메서드명 처럼 정말로 우승자인지 판단할 수 있나요?
2. 매개변수가 정말로 maxPosition만 입력될 수 있나요?

ulimy replied:
position의 일치 여부를 확인하는 함수이니 함수명은 isSamePosition으로, 매개변수명은 target으로 변경했습니다.

lalize commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,44 @@
+        return position;
+    }
+
+    public void moveOrNot(int randomNumber) {
```
randomNumber가 아닌 값도 입력될 수 있지 않나요?
Car의 입장에서 randomNumber가 의미가 있을지 한번 고민해보세요!

ulimy replied:
Car의 입장에서 매개변수는 random이든 아니든 별로 상관이 없군요. 어차피 기준값과의 차이만 확인하면 되니까요! 매개변수명 target으로 수정했습니다.

lalize commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,44 @@
+    public String toString() {
+        String positionSign = SIGN_OF_POSITION.repeat(position);
+        return (name + " : " + positionSign + "\n");
+    }
```
화면에 `-`가 아닌 `*`로 출력하고 싶을 때(화면 요구사항이 변경된 경우) 도메인단 변경이 필요하네요.
자동차의 상태를 알 수 있다면 View단에서 자유자재로 그릴 수 있지 않을까요?

ulimy replied:
출력문구 구성은 출력을 위한 기능이니 View에 선언되는게 맞는 것 같습니다! CarDTO를 활용하여 출력문구를 View에서 구성할 수 있도록 수정했습니다.

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,75 @@
+    public void initTotalAttempt(String attempt) {
+        validateAttempt(attempt);
+        totalAttempt = Integer.parseInt(attempt);
+    }
```
RacingGame이 생성된 후 해당 메서드를 호출하지 않고 시작(play)을 하면 어떤 문제가 발생할까요?
그리고 문제를 원천적으로 방어하려면 어떻게 해야할까요?

ulimy replied:
init 없이 play가 실행된다면 totalAttempt에 null값이 들어있게되므로 NullPointerException이 발생할 것 같습니다! 이를 피하기 위해 0으로 초기화 하도록 수정했습니다.

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,75 @@
+                .get()
+                .getPosition();
+
+    }
```
stream 사용 👍

그런데 아래 처럼 구현하면 더욱 가독성이 좋지 않을까요?

자바 stream api를 조금 더 학습해보시면 좋을 것 같아요.
자연스럽게 읽히는 코드를 더욱 쉽게 작성할 수 있어요!

```suggestion
        return cars.stream()
                .mapToInt(Car::getPosition)
                .max()
                .orElse(0);
    }
```

ulimy replied:
max를 이용하니 정렬할 필요가 없군요.. mapToInt로 position을 먼저 추출한 뒤 최댓값을 구했다면 getPosition을 한번 더 호출할 필요도 없구요..!  부끄럽습니다ㅜㅜ 더 공부하겠습니다.

lalize commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -0,0 +1,8 @@
+public class RandomUtil {
+
+    public static int generateRandomNumber() {
+        return (int) Math.floor(Math.random() * 10);
```
10이 어떤 의미를 갖는 걸까요?
매직넘버는 상수화 해주세요!

ulimy replied:
상수화 완료했습니다!

lalize commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -0,0 +1,8 @@
+package racingcar.util;
+
+public class RandomUtil {
```
다른 유틸 클래스처럼 인스턴스화를 막아 불필요한 인스턴스화를 피할 수 있도록 해봅시다~

ulimy replied:
놓쳤던 부분이네요ㅜㅜ 수정했습니다!

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,75 @@
+    private int totalAttempt;
+
+    public void initCarNames(String carNames) {
+        String[] names = carNames.split(",");
```
`,`는 어떤 의미를 가지고 있나요?
매직넘버는 상수화 해주세요!

ulimy replied:
상수화 완료했습니다!

lalize commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,75 @@
+        return totalExecutionResult.toString();
+    }
+
+    public List<String> selectWinners() {
```
👍

ulimy replied:
😊

lalize commented on src/main/java/racingcar/util/ValidationUtil.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private static void isEmpty(String carName) {
+        if (carName == null || carName.length() == 0) {
```
```suggestion
        if (carName == null || carName.isEmpty()) {
```

ulimy replied:
String에도 isEmpty가 있군요! 감사합니다.

lalize commented on src/main/java/racingcar/util/ValidationUtil.java:
```diff
@@ -0,0 +1,67 @@
+
+import static racingcar.constant.ErrorMessage.*;
+
+public class ValidationUtil {
```
ValidationUtil이란 이름처럼 유효성 검증을 위한 공통적인 부분만을 관리하도록 하면 좋지 않을까요?
도메인 유효성 체크에 대한 모든 로직들이 ValidationUtil에 오게 되면 이해하기 어렵고 객체의 응집도가 낮아지게 되지 않을까요?

ulimy replied:
응집도를 높이기 위해 각 도메인 유효성 체크를 서로 다른 파일에 구성되도록 수정했습니다. 공통적인 부분은 없어 ValidationUtil을 따로 두지는 않았습니다.

lalize commented on src/test/java/racingcar/game/RacingGameTest.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Test
+    @DisplayName("올바른 최종우승자 결정")
+    public void selectRightWinner() {
```
> 비즈니스 로직과 UI 로직 분리는 해보았으나 올바른 우승자를 선택하는지 확인하는 테스트 코드를 작성하지 못했습니다ㅠㅠ 제가 작성한 코드로는 어떻게 하면 좋을지 조언을 들을 수 있을까요? getCars 메서드를 선언하여 확인해보는 방법, 우승자 선별하는 코드를 다른 클래스로 분리하여 cars를 매개변수로 전달하도록 한 뒤 테스트 하는 방법 등을 떠올려보았으나 확신이 서지 않습니다..

오직 테스트를 위한 프로덕션 코드를 만드는 건 좋은 생각은 아닙니다.
현재 테스트를 어렵게 하는 메서드를 테스트 가능하도록 리팩토링하는 방향으로 고민해보세요.

위에 RacingGame 변경사항에 다음과 같은 코멘트 하나를 남겼는데요 `RacingGame이 생성된 후 해당 메서드를 호출하지 않고 시작(play)을 하면 어떤 문제가 발생할까요?` 이것에 대해 고민하여 해결해보시면 자연스럽게 RacingGame.selectWinners도 테스트 할 수 있을 거라고 봐요. 그래도 잘 모르겠다면 DM 주시면 도움 드릴게요 :)

ulimy replied:
이 부분이 제일 어려웠는데, DM으로 자세히 설명해주셔서 감사합니다. Car 생성자의 매개변수에  position을 추가함으로써 Car 도메인 관련 테스팅이 원활하도록 수정하였고, List를 활용하여 전체 실행 결과를 반환함으로써 최종우승자 관련 테스팅이 원활하도록 수정했습니다!

ulimy replied:
다시 생각해보니 position은 언제나 초기값이 0이므로 생성자의 매개변수에 position을 추가하는것은 적절하지 않는 것 같습니다. 매개변수에 position을 둠으로써 얻는 장점은 테스팅의 편리함 뿐이고, 테스팅을 위한 프로덕션 코드는 적절하지 않기 때문입니다. 따라서 초기값은 언제나 0으로 세팅하고, 테스트 코드를 수정하겠습니다!

lalize reviewed (changes requested):
안녕하세요 아리! 피드백 반영하느라 고생 많으셨어요 👍 
아쉬운 부분이 조금 있어 추가 코멘트 남겨드렸으니 충분히 고민하여 개선해보면 좋을 것 같아요.
다른 개선 포인트와 동일한 부분은 추가 코멘트 남기지 않은 부분도 있으니 찾아서 함께 개선해 주시면 감사하겠습니다!

ulimy committed:
style : 보다 명확한 의미를 갖는 이름으로 메서드명 수정

ulimy committed:
refactor : 코드 간결화

ulimy committed:
refactor : 클래스별 유효성 검증 코드 분리 및 테스트코드 수정

ulimy committed:
refactor : 자동차 이름 구분 기준 문자열 상수화

ulimy committed:
refactor : CarDTO 작성 및 테스트코드 수정

자동차의 진행상황 문구 구성을 Car가 아닌 OutputView에서 진행하도록 수정하기위해 CarDTO 작성, 수정된 소스코드를 위해  테스트코드 수정

ulimy committed:
refactor : Car의 생성자 수정 및 테스트코드 수정
RacingGame의 init()이 실행되지 않은채 play()가 실행될 경우 발생할 NullPointException을 방지 및 원활한 테스팅을 위해 Car의 생성자에서 position을 0으로 초기화 해주도록 수정, 수정된 소스코드를 위해 테스트코드 수정

ulimy committed:
test : 최종우승자 선정 테스트코드 작성

ulimy committed:
test : 보다 정확한 테스팅을 위해 조건 상세화

ulimy committed:
style : 통일성을 위해 매개변수명 target으로 통일

ulimy committed:
refactor : totalAttempt의 NullPointerException을 피하기위해 0으로 초기화

ulimy committed:
refactor : 생성자를 private으로 선언함으로써 인스턴스화 규제

ulimy committed:
refactor : 불필요한 Car의 생성자에 구현된 매개변수 position 제거 및 테스트코드 수정

lalize commented on src/main/java/racingcar/domain/CarDTO.java:
```diff
@@ -0,0 +1,19 @@
+package racingcar.domain;
+
+public class CarDTO {
```
불변 👍

ulimy replied:
😀😀

lalize commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,38 @@
+        return position;
+    }
+
+    public void moveOrNot(int target) {
```
자동차에게 이동하라 메시지만 보내고 이동할지 말지는 자동차 스스로 판단하여 이동한다고 생각했을 때
moveOrNot이라는 메서드명은 구현에 대한 내용이 조금 내포되어 있는 것 같아요. 아리는 어떻게 생각하시나요?
```suggestion
    public void move(int target) {
```

ulimy replied:
우선 제가 "moveOrNot" 이라는 메서드명을 사용한 이유부터 설명하겠습니다.
메서드명은 해당 메서드가 처리하는 일을 함축적으로 전달하는 것이라고 이해하고 있습니다.
따라서 메서드명에서 "자동차가 이동할지 말지 스스로 판단하여 이동시키는 함수" 라는 의미가 드러나기를 바랬습니다. 
하지만 단순히 move만을 사용한다면 그저 "자동차를 이동시키는 함수" 라는 의미만이 전달될 것 같습니다.

"구현에 대한 내용이 내포되어있는 것 같아요" 라는 문장을 조금만 더 자세히 설명해주실 수 있을까요?
메서드명에 구현에 관련된 내용이 내포되는 것이 적절하지 않다는 말씀이신지, 어떻게 이동시키든 저 메서드가 알아서 하는 것이니 "자동차를 이동시키는 함수"정도의 의미만 표현해도 충분하다는 말씀이신지, 혹은 제가 이해하지 못한  다른 뜻이 있는 것인지 궁금합니다!

lalize replied:
> 메서드명에 구현에 관련된 내용이 내포되는 것이 적절하지 않다는 말씀이신지, 어떻게 이동시키든 저 메서드가 알아서 하는 것이니 "자동차를 이동시키는 함수"정도의 의미만 표현해도 충분하다는 말씀이신지, 혹은 제가 이해하지 못한 다른 뜻이 있는 것인지 궁금합니다!

둘 다 해당 됩니다.

만약 요구사항이 변경되어 입력 range 마다 이동해야할 거리가 정해진 경우 각 케이스에 대한 모든 것을 메서드명으로 나타내야 할까요?

ulimy replied:
요구사항이 변경에 따라 메서드명이 변경 될 수 있다는 생각은 못했습니다.. move로 변경했습니다!

lalize commented on src/main/java/racingcar/domain/CarDTO.java:
```diff
@@ -0,0 +1,19 @@
+package racingcar.domain;
+
+public class CarDTO {
+    private final String carName;
```
중복인 것 같아요. name만으로 의미 전달이 되는 것 같습니다~

```suggestion
    private final String name;
```

ulimy replied:
동의합니다. 수정했습니다!

lalize commented on src/main/java/racingcar/util/CarNameValidationUtil.java:
```diff
@@ -0,0 +1,37 @@
+        isValidLength(carName, minLength, maxLength);
+    }
+
+    private static void isEmpty(String carName) {
```
is 접두어는 일반적으로 불린 타입 네이밍으로 쓰입니다.

아래 링크 참고하셔서 전체적으로 변수 네이밍 점검 부탁드릴게요!

https://tecoble.techcourse.co.kr/post/2020-04-24-variable_naming/

ulimy replied:
확인 후 수정했습니다.

lalize commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,54 @@
+        assertThat(car.getPosition()).isEqualTo(0);
+    }
+
+    @Test
+    @DisplayName("최대 전진횟수와 동일한 전진횟수를 가지는 경우 우승자임을 반환")
+    public void mustBeWinner() {
+        Car car = new Car("woo");
+
+        car.moveOrNot(9);
+        car.moveOrNot(9);
+        car.moveOrNot(9);
+
+        assertThat(car.isSamePosition(3)).isTrue();
+    }
+
+    @Test
+    @DisplayName("최대전진횟수와 동일하지 않는 전진횟수를 가지는 경우 우승자가 아님을 반환")
+    public void mustNotBeWinner() {
+        Car car = new Car("woo");
+
+        assertThat(car.isSamePosition(3)).isFalse();
+    }
```
테스트 대상이 실제로 우승자를 구하는 로직이 아닌 것 같아요.
해당 테스트는 의미 없는 테스트로 보입니다!

ulimy replied:
Car의 isSamePosition이 정상작동여부 확인을 위해 해당 테스트코드를 작성했습니다. 
또한 해당 테스트가 실제로 우승자를 구하는 로직이 아님에 동의합니다! 
그렇다면 의미 없는 테스트라고 느끼신 이유가 우승자를 구하는 로직이 아니기 때문인지 
혹은 isSamePosition이라는 메서드 자체가 테스트하기에 의미가 없기 때문인지 알 수 있을까요?

lalize replied:
테스트명과 테스트 코드가 잘못된 것 같아요.

isSamePosition이 우승자를 계산하는 기능인가요? 자동차 위치와 입력받은 위치가 같은지 판단하는 기능 아닌가요?

자동차 위치가 3으로 주어졌을 때 isSamePosition(3)를 호출하여 true가 반환되는지 검증해야 하지 않나요?

ulimy replied:
네 이해했습니다. 수정했습니다!

lalize commented on src/test/java/racingcar/game/RacingGameTest.java:
```diff
@@ -0,0 +1,55 @@
+
+public class RacingGameTest {
+
+    static RacingGame racingGame = new RacingGame();
+    static List<List<CarDTO>> totalExecutionResult;
+
+    @BeforeAll
+    public static void setRacingGame() {
+        racingGame.initCarNames("woo,te,co");
+        racingGame.initTotalAttempt("5");
+
+        totalExecutionResult = racingGame.play();
+    }
```
BeforeEach를 하는 게 낫지 않을까요?

```suggestion
    private RacingGame racingGame;
    private List<List<CarDTO>> totalExecutionResult;

    @BeforeEach
    public void setRacingGame() {
        racingGame = new RacingGame();
        racingGame.initCarNames("woo,te,co");
        racingGame.initTotalAttempt("5");
        
        totalExecutionResult = racingGame.play();
    }
```

ulimy replied:
해당 메서드는 전체 테스트 전 단 한번만 실행되어도 충분할 것 같아 BeforeAll을 사용했습니다!

lalize commented on src/test/java/racingcar/game/RacingGameTest.java:
```diff
@@ -0,0 +1,55 @@
+        assertThat(racingGame.selectWinners()).isEqualTo(winners);
+    }
+
+    private int getMaxPosition(List<CarDTO> cars) {
+        return cars.stream()
+                .mapToInt(CarDTO::getPosition)
+                .max()
+                .orElse(0);
+    }
+
+    private List<String> selectWinners(List<CarDTO> cars, int targetNum) {
+        return cars.stream()
+                .filter(c -> c.getPosition() == targetNum)
+                .map(CarDTO::getCarName)
+                .collect(Collectors.toList());
+    }
```
테스트를 위해 비즈니스 로직이 테스트로 내려온 것 같아요.
해당 로직이 잘못되어 있으면 버그 발견이 어렵지 않나요?

자동차를 생성하여 수동으로 움직이는 코드를 작성하면 우승자가 누군지 알 수 있을 거예요.
해당 값과 racingGame.selectWinners()를 비교하도록 테스트를 작성해보세요.

lalize commented on src/main/java/racingcar/util/CarNameValidationUtil.java:
```diff
@@ -0,0 +1,37 @@
+    private CarNameValidationUtil() {
+    }
+
+    public static void validateCarName(String carName, int minLength, int maxLength) {
```
자동차 생성 책임으로 유효성 체크를 할 순 없을까요?
모든 도메인 객체 생성의 유효성 체크를 각각 다른 클래스로 관리하면 비용이 많이 들지 않을까요?

ulimy replied:
저도 이 부분에 대해서 항상 고민하게 되는데요, 한 곳에서 모두 관리한다면 코드의 길이가 너무 길어지게 되어 가독성이 떨어지지 않을까 생각했습니다. 코드 길이와 클래스 분리에 대한 핀의 의견을 여쭤봐도 될까요?

lalize replied:
도메인 객체를 구성하는 더 작은 도메인 객체를 만들 수도 있겠네요!


참고
* https://tecoble.techcourse.co.kr/post/2020-05-29-wrap-primitive-type/

ulimy replied:
와 너무 좋은 방법이네요! 알려주셔서 감사합니다 핀 :)

lalize commented on src/main/java/racingcar/constant/ErrorMessage.java:
```diff
@@ -0,0 +1,18 @@
+package racingcar.constant;
+
+public class ErrorMessage {
```
이것도 실제로 사용하는 곳에서 선언하면 좋을 것 같네요.

ulimy replied:
네 동의합니다! 수정했습니다!

lalize commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,38 @@
+    private final int MAX_LENGTH_OF_NAME = 5;
+    private final int CONDITION_FOR_MOVE = 4;
+
+    public Car(String name) {
+        validateCarName(name, MIN_LENGTH_OF_NAME, MAX_LENGTH_OF_NAME);
+        this.name = name;
+        this.position = 0;
+    }
```
완전한 객체를 생성하는 방법이 많으면 많을수록 사용하기 좋은 객체라고 생각합니다.
모든 멤버변수 초기화 값을 입력받는 생성자를 만드는 게 꼭 테스트를 위한 거라고 생각되진 않아요.

lalize reviewed (approved):
안녕하세요 아리 🐣 리뷰 반영 잘 하셨어요! 💯
1단계는 여기서 머지하겠습니다! 고생하셨어요~!
추가 코멘트 남겼으니 다음 단계에서 반영해주시면 좋을 것 같아요~~
